### ISSUE TITLE
Json / Yaml config files

### ISSUE THREAD

**Issue by b-studios** [scallop/scallop#218]
Are there any plans to support other formats of config files? The existing way to read configs from a file is already a good start, but doesn't scale so well to larger configs.

**Comment by Rogach** [comment]
I didn't have such plans, but the idea seems reasonable.

Do you have suggestions on how such JSON/YAML file should look like?

Also I'm not sure how to handle the detection of file format if reading options from stdin. When we are reading from file we can use file extensions, but when reading from stdin we don't have filenames and thus need to rely on contents.

**Comment by b-studios** [comment]
Thanks for answering @Rogach. I am using scallop in our implementation of the [Effekt language](https://effekt-lang.org). Right now, we only allow passing configuration as additional arguments, but realized that some kind of config file / project description etc. would be useful. 

I haven't thought really long about it, but one could easily map types like strings, lists, numbers, and filepath to some version in json (or any other markup language). For instance, in Effekt, we have the following configurations:

```scala
val compile: ScallopOption[Boolean] = toggle(
    "compile",
    descrYes = "Compile the Effekt program",
    descrNo = "Run the effekt program in the interpreter",
    default = Some(false)
)

val outputPath: ScallopOption[File] = opt[File](
    "out",
    descr = "Path to write generated files to (defaults to ./out)",
    default = Some(new File("./out")),
    required = false
)

val includePath: ScallopOption[List[File]] = opt[List[File]](
    "includes",
    descr = "Path to consider for includes (can be set multiple times)",
    default = Some(List(new File("."))),
    noshort = true
)
```

a json file could look like:

```json
{
  "compile": true,
  "out": "./build",
  "includes": [
    "lib",
    "src"
  ]
}
```

or in yaml:
```yaml
compile: true
out: "./build"
includes:
  - "lib"
  - "src"
```

**Comment by b-studios** [comment]
Regarding detection: It would be good enough to leave this to the developers. 
As a library author, I am ok with deciding myself how to detect a config file. This could be based on convention, as an additional cmd line argument, as a command line option, a combination, etc.

**Comment by Rogach** [comment]
I think this is doable, but I won't be able to implement it quickly.

At first I thought it could be done by just blindly converting JSON object into string arguments - e.g. your example would become `Seq("--compile", "true", "--out", "./build", "--includes", "lib", "src")` - and passing it into the same parser.

But I think this approach is very incomplete - we lose information by converting everything to strings, and there can be cases when such approach will lead to incorrect argument assignment in multi-arg options. Plus sucommands will be parsed incorrectly, and trailing args won't work at all. Toggle, tally or property options won't work with such approach as well.

So I think much more careful approach is required - will need to completely bypass CLI argument parsing pipeline for such cases.

Another important question is how to handle the initial file parsing. I do not want to implement the parsers myself (I think that in theory I can do it, but the result would probably be untested, buggy and incomplete). And I'd like to avoid adding hard dependencies on other libraries if at all possible. Maybe such functionality should reside in a separate subproject?

For reference, here's an example of "naive" approach ("just flatten the config"):

```scala
import org.rogach.scallop._
import play.api.libs.json._
import java.nio.file.{Files, Paths}

class Conf(args: Seq[String]) extends ScallopConf(args) {
  val logLevel = opt[String]()
  val optimizationLevel = opt[Int]()
  object compile extends Subcommand("compile") {
    val outputPath = opt[String]("out")
    val includePath = opt[List[String]]("include")
  }
  addSubcommand(compile)
  verify()
}

object Main {
  def main(args: Array[String]): Unit = {
    val conf =
      args.headOption match {
        case Some(arg) if arg.startsWith("@") && arg.endsWith(".json") =>
          val path = arg.stripPrefix("@")
          val content = new String(Files.readAllBytes(Paths.get(path)), "UTF-8")
          new Conf(flattenConfig(Json.parse(content).as[JsObject]))
        case _ =>
          new Conf(args.toSeq)
      }

    println(conf.logLevel())
    println(conf.optimizationLevel())
    println(conf.compile.outputPath())
    println(conf.compile.includePath())
  }

  private def flattenConfig(obj: JsObject): Seq[String] = {
    obj.fields.toSeq.flatMap { case (key, value) =>
      value match {
        case v: JsObject => key +: flattenConfig(v)
        case a: JsArray => ("--" + key) +: a.value.map(_.toString)
        case other => Seq(("--" + key), other.toString)
      }
    }
  }
}
```

**Comment by b-studios** [comment]
> I think this is doable, but I won't be able to implement it quickly.

Don't worry. I just wanted to bring it up, since I believe in the long run this would be very valuable.

> At first I thought it could be done by just blindly converting JSON object into string arguments

I wouldn't start going down this road as it will introduce more problems than it solves :)

> will need to completely bypass CLI argument parsing pipeline for such cases.

Is it possible to properly parse a config file and then merge the config with the ones provided at the command line? Trailing args should not be possible with config files, I agree.

> separate subproject?

Yes, that's ok. Also to reduce dependencies on the original project, if not used. 

In fact, I would prefer leaving parsing to the library users. Probably a convenient (and somewhat type-safe) programmatic way to set configuration without generating and reparsing argument strings would suffice as an interface.

**Comment by Rogach** [comment]
> Probably a convenient (and somewhat type-safe) programmatic way to set configuration without generating and reparsing argument strings would suffice as an interface.

One simple way to use such approach would be to just pass the config values as a default arguments (it should work even with current Scallop version). For example:

```scala
val fileConfig: FileConfig = readConfig(args)
class Conf(config: FileConfig, args: Seq[String]) extends ScallopConf(args) {
  val outputFile = opt[File]("out", default = config.outputFile)
  verify()
}
val config = new Conf(fileConfig, args)
```

Not very convenient (duplicate effort for each option), but reasonably type-safe.

The other approach - setting configuration after CLI parsing is done - is a bit convoluted because we get a somewhat circular relationship - we can't verify the args before loading the config but user can't load the config before we verified the args. For example if `outputFile` is required, then `ScallopConf.verify` would throw an error and user-code would never get a chance to insert a value from config file.

> Trailing args should not be possible with config files, I agree.

They shouldn't be possible with "naive" approach, but with full-blown config parsing this should be easy - trailing arguments still have a name, thus we have a key in config file from which to extract the values.

To solve the dependency problem I think some API like this may be doable:

```scala
class Conf(args: Seq[String]) extends ScallopConf(args) {
  registerConfigFileReader(ScallopJSONReader)
  // ScallopJSONReader extends ScallopConfigFileReader,
  // and resides in a separate module to move extra dependencies into a separate module
  // ... usual option definitions
}
// config file reader interface shouldn't be very complex,
// so library users would be implement their own if needed
trait ScallopConfigFileReader[T] { // T is used to signify the "base" config object type
  def isFilenameSupported(filename: String) // here we can just check for file extension, or actually check the file contents
  def readConfigObject(filename: String): T
  def getOptionValue(obj: T, name: String, converter: ValueConverter[A]): A // still not sure how to handle the individual value type checking, but ValueConverter (or something similar) is definitely needed here in order to pass the item type
  def getSubcommandObject(obj: T, subcommandName: String): T // or subcommandNames: Seq[String], to support aliases
}
```

This way we can separate different formats into different subprojects and it should be easy for a user to just implement his own if need arises (the code for ScallopConfigFileReader should probably fit into 20-30 lines at most - an easy price for avoiding loading in extra JSON library).

**Comment by b-studios** [comment]
> the code for ScallopConfigFileReader should probably fit into 20-30 lines at most - an easy price for avoiding loading in extra JSON library

Yes! Something like this sounds great. However, your proposed workaround with defaults also already sounds like a first viable option. 

Thanks for taking your time to think about this.
